backend/clickhouse/sessions.go

```diff
@@ -48,36 +74,89 @@ func (client *Client) WriteSessions(ctx context.Context, sessions []*model.Sessi
+			chf := ClickhouseField{
+				ProjectID: int32(session.ProjectID),
+				Type:      field.Type,
+				Name:      field.Name,
+				Value:     field.Value,
+				SessionID: int64(session.ID),
+			}
+			chFields = append(chFields, &chf)
 		}
 
-		viewedByAdmins := []int32{}
+		var viewedByAdmins clickhouse.ArraySet
 		for _, admin := range session.ViewedByAdmins {
 			viewedByAdmins = append(viewedByAdmins, int32(admin.ID))
 		}
 
-		ch := ClickhouseSession{
-			Session:        *session,
-			ID:             int64(session.ID),
-			Fingerprint:    int32(session.Fingerprint),
-			ProjectID:      int32(session.ProjectID),
-			PagesVisited:   int32(session.PagesVisited),
-			ViewedByAdmins: viewedByAdmins,
-			FieldKeys:      fieldKeys,
-			FieldKeyValues: fieldKeyValues,
+		chs := ClickhouseSession{
+			ID:                 int64(session.ID),
+			Fingerprint:        int32(session.Fingerprint),
+			ProjectID:          int32(session.ProjectID),
+			PagesVisited:       int32(session.PagesVisited),
+			ViewedByAdmins:     viewedByAdmins,
+			FieldKeys:          fieldKeys,
+			FieldKeyValues:     fieldKeyValues,
+			CreatedAt:          session.CreatedAt,
+			UpdatedAt:          session.UpdatedAt,
+			SecureID:           session.SecureID,
+			Identified:         session.Identified,
+			Identifier:         session.Identifier,
+			City:               session.City,
+			Country:            session.Country,
+			OSName:             session.OSName,
+			OSVersion:          session.OSVersion,
+			BrowserName:        session.BrowserName,
+			BrowserVersion:     session.BrowserVersion,
+			Processed:          session.Processed,
+			HasRageClicks:      session.HasRageClicks,
+			HasErrors:          session.HasErrors,
+			Length:             session.Length,
+			ActiveLength:       session.ActiveLength,
+			Environment:        session.Environment,
+			AppVersion:         session.AppVersion,
+			FirstTime:          session.FirstTime,
+			Viewed:             session.Viewed,
+			WithinBillingQuota: session.WithinBillingQuota,
+			EventCounts:        session.EventCounts,
+			Excluded:           session.Excluded,
+			Normalness:         session.Normalness,
 		}
 
-		if err := batch.AppendStruct(&ch); err != nil {
-			return err
-		}
+		chSessions = append(chSessions, &chs)
 	}
 
-	return batch.Send()
+	chCtx := clickhouse.Context(ctx, clickhouse.WithSettings(clickhouse.Settings{
+		"async_insert":          1,
+		"wait_for_async_insert": 1,
+	}))
+
+	fieldsSql, fieldsArgs := sqlbuilder.
+		NewStruct(new(ClickhouseField)).
+		InsertInto(FieldsTable, chFields...).
+		BuildWithFlavor(sqlbuilder.ClickHouse)
+
+	sessionsSql, sessionsArgs := sqlbuilder.
+		NewStruct(new(ClickhouseSession)).
+		InsertInto(SessionsTable, chSessions...).
+		BuildWithFlavor(sqlbuilder.ClickHouse)
+
+	var g errgroup.Group
+	g.Go(func() error {
+		return client.conn.Exec(chCtx, fieldsSql, fieldsArgs...)
+	})
+
+	g.Go(func() error {
+		return client.conn.Exec(chCtx, sessionsSql, sessionsArgs...)
+	})
```

Comment:
neat! just checking, the context and connection are thread-safe?

Reply:
The context here should be as it's read only. I believe `client.conn` is as well (e.g. there's a `MaxOpenConns` option https://github.com/ClickHouse/clickhouse-go/blob/main/clickhouse_options.go#L137) but not 100% sure on that (I was seeing deadlocks before when both Logs and Sessions were writing in batches), will monitor this after deploying and revert if seeing the same deadlocks

Reply:
makes sense. should we try increase the `MaxOpenConns` (defaults to 10) so that the 32 workers per ecs node can make more concurrent writes?
https://clickhouse.com/docs/en/integrations/go#connection-pooling

Reply:
I think because the `flush`es for the batched topic are surrounded with a mutex, there won't be more than 4 connections at a time (writing logs, traces, sessions, fields)
https://sourcegraph.com/github.com/highlight/highlight/-/blob/backend/worker/kafka_worker.go?L397
also observed from rewinding the queue on Friday that most of the time is spent waiting for async inserts to complete, I don't know if it's worth having multiple workers just to read from Kafka since they I think they spend most of their time waiting for this mutex

Reply:
cool, good to know. thanks for looking into that. sounds like removing the mutex and writing more concurrently is a bad idea overall since we'd just be increasing that max parts value, maybe to a point that clickhouse won't be happy.